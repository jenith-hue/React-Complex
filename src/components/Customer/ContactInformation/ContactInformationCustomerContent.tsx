import {
  RACButton,
  RACTable,
  RACTableCell,
  RACTableRow,
  RACTooltip,
  Typography,
} from '@rentacenter/racstrap';

import React, { useState } from 'react';
import { useLocation } from 'react-router-dom';
import { useCustomerDetails } from '../../../context/CustomerDetails/CustomerDetailsProvider';
import { CustomerLocationState, Phone } from '../../../types/types';
import { EditContactInformationModal } from './EditContactInformationModal';
import { withContactStyles, WrappedComponentProps } from './withContactStyles';
import { abbrev, formatPhoneNumber } from '../../../utils/utils';
import clsx from 'clsx';
import { useTextConversationActions } from '../../../context/TextConversationProvider/TextConversationProvider';

const ContactInformationCustomerContent_ = ({
  classes,
  onLogClicked,
}: WrappedComponentProps) => {
  const {
    customerDetails: { phones, firstName, lastName },
    isLogActivityAllowed,
  } = useCustomerDetails();
  const location = useLocation<CustomerLocationState>();

  const [
    showEditCustomerInformationModal,
    setShowEditCustomerInformationModal,
  ] = useState<boolean>(false);

  const [tempPhone, setTempPhone] = useState<Phone>();
  const popUpPhoneInstructions = (phone: Phone): void => {
    setTempPhone(phone);
    setShowEditCustomerInformationModal(true);
  };

  const closePhoneInstructionModal = () => {
    setShowEditCustomerInformationModal(false);
  };
  const { updatePhoneOptionsfromEdit } = useTextConversationActions();
  const updatePhoneNotes = (
    phoneId: string,
    notes: string,
    bestTimeToCall: string,
    phoneNumber: string,
    phoneType: string,
    ext: string
  ) => {
    phones?.map((one) => {
      if (phoneId == one.phoneId) {
        one.note = notes;
        one.callTimeTypeDesc = bestTimeToCall;
        one.phoneNumber = phoneNumber;
        one.extension = ext;
        one.phoneTypeDesc = phoneType;
      }
    });
    if (phoneType === 'Cell Phone') {
      updatePhoneOptionsfromEdit();
    }
  };
  const renderTableHead = () => (
    <>
      <RACTableCell>Name</RACTableCell>
      <RACTableCell>Phone Number</RACTableCell>
      <RACTableCell>Type</RACTableCell>
      <RACTableCell>Best Time to Call</RACTableCell>
      <RACTableCell>Phone Instructions</RACTableCell>
      <RACTableCell># Contacts Today</RACTableCell>
      <RACTableCell>Last Call Result</RACTableCell>
      <RACTableCell align="center">Action</RACTableCell>
    </>
  );

  const customerId =
    location?.state?.customer?.customerId || location?.pathname?.split('/')[3];

  const renderTableContent = () => (
    <>
      {phones?.map((phone) => (
        <RACTableRow backgroundColor="white" key={phone.phoneId}>
          <RACTableCell>
            <Typography variant="body2" className={classes.tableContentColor}>
              {`${firstName} ${lastName}`}
            </Typography>
          </RACTableCell>
          <RACTableCell>
            <Typography variant="body2" className={classes.tableContentColor}>
              {formatPhoneNumber(
                phone.phoneNumber?.split('.')[0],
                !isLogActivityAllowed
              )}
            </Typography>
          </RACTableCell>
          <RACTableCell>
            <Typography variant="body2" className={classes.tableContentColor}>
              {phone.phoneTypeDesc}
            </Typography>
          </RACTableCell>
          <RACTableCell>
            <Typography variant="body2" className={classes.tableContentColor}>
              {phone.callTimeTypeDesc}
            </Typography>
          </RACTableCell>
          <RACTableCell>
            <RACTooltip placement="right-start" title={phone?.note || ''}>
              <Typography
                noWrap
                variant="body2"
                color="primary"
                className={clsx(phone.note)}
              >
                <span>{phone.note ? abbrev(phone.note) : ''}</span>
              </Typography>
            </RACTooltip>
            <Typography noWrap variant="body2" color="primary"></Typography>
          </RACTableCell>
          <RACTableCell>
            <Typography
              align="center"
              variant="body2"
              className={classes.tableContentColor}
            >
              {phone.communicationsToday}
            </Typography>
          </RACTableCell>
          <RACTableCell>
            <RACTooltip
              placement="right-start"
              title={phone.lastCallResultDescription || ''}
            >
              <Typography noWrap variant="body2" color="primary">
                <span>{abbrev(phone.lastCallResultDescription || '')}</span>
              </Typography>
            </RACTooltip>
          </RACTableCell>
          <RACTableCell className={classes.center}>
            <RACButton
              variant="outlined"
              color="primary"
              className={classes.logButton}
              onClick={() =>
                onLogClicked(customerId, phone.phoneNumber?.split('.')[0])
              }
              disabled={!phone?.phoneNumber || !isLogActivityAllowed}
            >
              Log
            </RACButton>
            <RACButton
              variant="contained"
              color="primary"
              className={classes.editButton}
              onClick={() => popUpPhoneInstructions(phone)}
            >
              Edit
            </RACButton>
          </RACTableCell>
        </RACTableRow>
      ))}
    </>
  );
  return (
    <>
      <EditContactInformationModal
        open={showEditCustomerInformationModal}
        phoneId={tempPhone?.phoneId || ''}
        note={tempPhone?.note || ''}
        ext={tempPhone?.extension || ''}
        bestTimeToCall={tempPhone?.callTimeType || ''}
        phoneType={tempPhone?.phoneType || ''}
        phoneNumber={tempPhone?.phoneNumber || ''}
        onChange={updatePhoneNotes}
        onClose={closePhoneInstructionModal}
      />
      <RACTable
        renderTableHead={renderTableHead}
        renderTableContent={renderTableContent}
      />
    </>
  );
};

export const ContactInformationCustomerContent = (props: any) =>
  withContactStyles(props)(ContactInformationCustomerContent_);
