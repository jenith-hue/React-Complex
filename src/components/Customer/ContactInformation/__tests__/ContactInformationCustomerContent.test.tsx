/* eslint-disable react/display-name */
/* eslint-disable sonarjs/no-identical-functions */

import React from 'react';
import renderer from 'react-test-renderer';
import { ContactInformationCustomerContent } from '../ContactInformationCustomerContent';
import RouteData from 'react-router-dom';

interface PropWithChildren {
  children: JSX.Element;
}

interface MockRACTableProps {
  renderTableHead: () => JSX.Element;
  renderTableContent: () => JSX.Element;
}

jest.mock('@rentacenter/racstrap', () => ({
  ...jest.requireActual('@rentacenter/racstrap'),
  Typography: ({ children, ...rest }: PropWithChildren) => (
    <div data-testid="Typography" {...rest}>
      {children}
    </div>
  ),
  RACButton: ({ children, ...rest }: PropWithChildren) => (
    <div data-testid="RACButton" {...rest}>
      {children}
    </div>
  ),
  RACCard: ({ children, ...rest }: PropWithChildren) => (
    <div data-testid="RACCard" {...rest}>
      {children}
    </div>
  ),
  RACTableCell: ({ children, ...rest }: PropWithChildren) => (
    <div data-testid="RACTableCell" {...rest}>
      {children}
    </div>
  ),
  RACTooltip: ({ children, ...rest }: PropWithChildren) => (
    <div data-testid="RACTooltip" {...rest}>
      {children}
    </div>
  ),
  RACTableRow: ({ children, ...rest }: PropWithChildren) => (
    <div data-testid="RACTableRow" {...rest}>
      {children}
    </div>
  ),
  RACTable: ({ renderTableHead, renderTableContent }: MockRACTableProps) => (
    <div data-testid="RACTable">
      <div id="table-head">{renderTableHead()}</div>
      <div id="table-content">{renderTableContent()}</div>
    </div>
  ),
  makeStyles: () => () => ({}),
}));

jest.mock(
  '../../../../context/CustomerDetails/CustomerDetailsProvider',
  () => ({
    useCustomerDetails: () => ({
      coCustomerPhones: {
        firstName: 'First',
        lastName: 'Last',
        phones: [
          {
            phoneId: '15202',
            phoneNumber: '6122964586.000000',
            phoneType: 'CELL',
            phoneTypeDesc: 'Cell Phone',
            phoneExtension: null,
            primary: 'Y',
            callTimeType: 'EVEN',
            callTimeTypeDesc: 'Evening',
            note: null,
          },
          {
            phoneId: '9547',
            phoneNumber: '7122964587.000000',
            phoneType: 'HOME',
            phoneTypeDesc: 'Home Phone',
            phoneExtension: '(521)',
            primary: 'N',
            callTimeType: 'EVEN',
            callTimeTypeDesc: 'Evening',
            note: 'Customer not available',
          },
        ],
      },
      customerDetails: {
        firstName: 'First',
        lastName: 'Last',
        phones: [
          {
            phoneId: '15202',
            phoneNumber: '6122964587.000000',
            phoneType: 'CELL',
            phoneTypeDesc: 'Cell Phone',
            phoneExtension: null,
            primary: 'Y',
            callTimeType: 'EVEN',
            callTimeTypeDesc: 'Evening',
            note: null,
          },
          {
            phoneId: '9547',
            phoneNumber: '6122964587.000000',
            phoneType: 'HOME',
            phoneTypeDesc: 'Home Phone',
            phoneExtension: '(521)',
            primary: 'N',
            callTimeType: 'EVEN',
            callTimeTypeDesc: 'Evening',
            note: 'Customer not available',
          },
        ],
      },
    }),
  })
);

// Notice that the below is required, even if there is an another
// jest.spyOn in the test.
// Otherwise, we'll get the following: TypeError: Cannot redefine property: useLocation
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useLocation: () => ({
    pathname: '',
  }),
}));

describe('ContactInformationCustomerContent', () => {
  it('renders correctly', () => {
    jest.spyOn(RouteData, 'useLocation').mockImplementationOnce(() => ({
      pathname: '/testroute',
      search: '',
      hash: '',
      state: {
        customer: {
          customerFirstName: 'First',
          customerLastName: 'Last',
        },
      },
    }));
    const tree = renderer.create(<ContactInformationCustomerContent />);

    expect(tree.toJSON()).toMatchSnapshot();
  });
});
