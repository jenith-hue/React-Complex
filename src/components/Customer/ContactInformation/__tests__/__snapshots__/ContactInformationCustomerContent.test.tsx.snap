// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactInformationCustomerContent renders correctly 1`] = `
<div
  data-testid="RACTable"
>
  <div
    id="table-head"
  >
    <div
      data-testid="RACTableCell"
    >
      Name
    </div>
    <div
      data-testid="RACTableCell"
    >
      Phone Number
    </div>
    <div
      data-testid="RACTableCell"
    >
      Type
    </div>
    <div
      data-testid="RACTableCell"
    >
      Best Time to Call
    </div>
    <div
      data-testid="RACTableCell"
    >
      Phone Instructions
    </div>
    <div
      data-testid="RACTableCell"
    >
      # Contacts Today
    </div>
    <div
      data-testid="RACTableCell"
    >
      Last Call Result
    </div>
    <div
      align="center"
      data-testid="RACTableCell"
    >
      Action
    </div>
  </div>
  <div
    id="table-content"
  >
    <div
      backgroundColor="white"
      data-testid="RACTableRow"
    >
      <div
        data-testid="RACTableCell"
      >
        <div
          data-testid="Typography"
          variant="body2"
        >
          First Last
        </div>
      </div>
      <div
        data-testid="RACTableCell"
      >
        <div
          data-testid="Typography"
          variant="body2"
        >
          (***) ***-****
        </div>
      </div>
      <div
        data-testid="RACTableCell"
      >
        <div
          data-testid="Typography"
          variant="body2"
        >
          Cell Phone
        </div>
      </div>
      <div
        data-testid="RACTableCell"
      >
        <div
          data-testid="Typography"
          variant="body2"
        >
          Evening
        </div>
      </div>
      <div
        data-testid="RACTableCell"
      >
        <div
          data-testid="RACTooltip"
          placement="right-start"
          title=""
        >
          <div
            className=""
            color="primary"
            data-testid="Typography"
            noWrap={true}
            variant="body2"
          >
            <span>
              
            </span>
          </div>
        </div>
        <div
          color="primary"
          data-testid="Typography"
          noWrap={true}
          variant="body2"
        />
      </div>
      <div
        data-testid="RACTableCell"
      >
        <div
          align="center"
          data-testid="Typography"
          variant="body2"
        />
      </div>
      <div
        data-testid="RACTableCell"
      >
        <div
          data-testid="RACTooltip"
          placement="right-start"
          title=""
        >
          <div
            color="primary"
            data-testid="Typography"
            noWrap={true}
            variant="body2"
          >
            <span>
              
            </span>
          </div>
        </div>
      </div>
      <div
        data-testid="RACTableCell"
      >
        <div
          color="primary"
          data-testid="RACButton"
          disabled={true}
          onClick={[Function]}
          variant="outlined"
        >
          Log
        </div>
        <div
          color="primary"
          data-testid="RACButton"
          onClick={[Function]}
          variant="contained"
        >
          Edit
        </div>
      </div>
    </div>
    <div
      backgroundColor="white"
      data-testid="RACTableRow"
    >
      <div
        data-testid="RACTableCell"
      >
        <div
          data-testid="Typography"
          variant="body2"
        >
          First Last
        </div>
      </div>
      <div
        data-testid="RACTableCell"
      >
        <div
          data-testid="Typography"
          variant="body2"
        >
          (***) ***-****
        </div>
      </div>
      <div
        data-testid="RACTableCell"
      >
        <div
          data-testid="Typography"
          variant="body2"
        >
          Home Phone
        </div>
      </div>
      <div
        data-testid="RACTableCell"
      >
        <div
          data-testid="Typography"
          variant="body2"
        >
          Evening
        </div>
      </div>
      <div
        data-testid="RACTableCell"
      >
        <div
          data-testid="RACTooltip"
          placement="right-start"
          title="Customer not available"
        >
          <div
            className="Customer not available"
            color="primary"
            data-testid="Typography"
            noWrap={true}
            variant="body2"
          >
            <span>
              Customer not ...
            </span>
          </div>
        </div>
        <div
          color="primary"
          data-testid="Typography"
          noWrap={true}
          variant="body2"
        />
      </div>
      <div
        data-testid="RACTableCell"
      >
        <div
          align="center"
          data-testid="Typography"
          variant="body2"
        />
      </div>
      <div
        data-testid="RACTableCell"
      >
        <div
          data-testid="RACTooltip"
          placement="right-start"
          title=""
        >
          <div
            color="primary"
            data-testid="Typography"
            noWrap={true}
            variant="body2"
          >
            <span>
              
            </span>
          </div>
        </div>
      </div>
      <div
        data-testid="RACTableCell"
      >
        <div
          color="primary"
          data-testid="RACButton"
          disabled={true}
          onClick={[Function]}
          variant="outlined"
        >
          Log
        </div>
        <div
          color="primary"
          data-testid="RACButton"
          onClick={[Function]}
          variant="contained"
        >
          Edit
        </div>
      </div>
    </div>
  </div>
</div>
`;
